In [34]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

In [35]:
from google.colab import files

uploaded = files.upload()

import pandas as pd

df = pd.read_csv('creditcard.csv')

df.head()

Saving creditcard.csv to creditcard (1).csv


Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [39]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [40]:
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

In [41]:
legit = df[df.Class == 0]
fraud = df[df.Class == 1]

In [42]:
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [43]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [44]:
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

In [45]:
legit_sample = legit.sample(n=492)

In [46]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [47]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
261294  159950.0  2.059049 -0.091209 -1.104368  0.396305 -0.116858 -1.099415   
2838      2417.0  1.326259 -0.666770 -0.169475 -0.910091 -0.535406 -0.363619   
134302   80742.0  1.255439  0.307729  0.292700  0.699873 -0.428876 -1.088456   
66585    52121.0  1.458303 -0.985494  0.415150 -1.369877 -1.323025 -0.537783   
150332   93077.0  2.069777  0.154763 -1.154613  0.558069  0.126579 -1.223059   

              V7        V8        V9  ...       V21       V22       V23  \
261294  0.138076 -0.279506  0.611536  ... -0.293384 -0.724221  0.343958   
2838   -0.407938 -0.104308 -1.102720  ...  0.245423  0.477835 -0.281008   
134302  0.043840 -0.167739  0.128854  ... -0.294795 -0.882126  0.136846   
66585  -0.937896 -0.144694 -1.933391  ... -0.174688 -0.069600  0.049698   
150332  0.194659 -0.498293  1.781696  ...  0.107594  0.766872  0.011083   

             V24       V25       V26       V27       V28  Amount  Class  
261294 -0.095863 -0.318753  0.206960 -0.070434 -0.062244    1.29      0  
2838   -0.459678  0.688523 -0.088771 -0.022213  0.002241   70.80      0  
134302  0.327949  0.194459  0.096516 -0.027271  0.029491    1.98      0  
66585   0.079148  0.329734 -0.179941  0.048154  0.020524   23.23      0  
150332 -0.019098  0.252306 -0.146233 -0.038786 -0.064314    1.00      0  

[5 rows x 31 columns]

In [48]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [49]:
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [50]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      95996.182927 -0.008389 -0.051706 -0.059565  0.008511  0.073661   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.003665  0.034658  0.047671  0.055614  ...  0.045180  0.021880   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.002988 -0.022007 -0.024343 -0.026861 -0.008222 -0.033293  0.011490   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       96.983577  
1      122.211321  

[2 rows x 30 columns]

In [51]:
X= new_dataset.drop(columns='Class', axis=1)
y= new_dataset['Class']

In [52]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
261294  159950.0  2.059049 -0.091209 -1.104368  0.396305 -0.116858 -1.099415   
2838      2417.0  1.326259 -0.666770 -0.169475 -0.910091 -0.535406 -0.363619   
134302   80742.0  1.255439  0.307729  0.292700  0.699873 -0.428876 -1.088456   
66585    52121.0  1.458303 -0.985494  0.415150 -1.369877 -1.323025 -0.537783   
150332   93077.0  2.069777  0.154763 -1.154613  0.558069  0.126579 -1.223059   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [53]:
print(y)

261294    0
2838      0
134302    0
66585     0
150332    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


In [54]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=2)

In [55]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


In [90]:
#creating a model

import torch
from torch import nn

class NeuralNetwork(nn.Module):
  def __init__(self, input_features, output_featues, hidden_units):
    super().__init__()
    self.linear_layer_stack = nn.Sequential(
        nn.Linear(in_features=input_features, out_features=hidden_units),
        nn.ReLU(),
        nn.Linear(in_features=hidden_units, out_features=hidden_units),
        nn.ReLU(),
        nn.Linear(in_features=hidden_units, out_features=hidden_units), # Added another hidden layer
        nn.ReLU(),
        nn.Linear(in_features=hidden_units, out_features=output_featues)
    )

  def forward(self, x):
    return self.linear_layer_stack(x)

model = NeuralNetwork(30, 1, 256)

model

NeuralNetwork(
  (linear_layer_stack): Sequential(
    (0): Linear(in_features=30, out_features=256, bias=True)
    (1): ReLU()
    (2): Linear(in_features=256, out_features=256, bias=True)
    (3): ReLU()
    (4): Linear(in_features=256, out_features=256, bias=True)
    (5): ReLU()
    (6): Linear(in_features=256, out_features=1, bias=True)
  )
)

In [91]:
#creating loss, optimizer, accuracy function

loss_fn = nn.BCEWithLogitsLoss()
optimizer = torch.optim.Adam(params=model.parameters(), lr=0.005) # Increased learning rate

def accuracy_fn(y_true, y_pred):
  correct = torch.eq(y_true, y_pred).sum().item()
  acc = (correct / len(y_pred)) * 100
  return acc

In [92]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

# Fit and transform 'Time' and 'Amount'
df['Time'] = scaler.fit_transform(df['Time'].values.reshape(-1, 1))
df['Amount'] = scaler.fit_transform(df['Amount'].values.reshape(-1, 1))

# Re-create the balanced dataset with scaled 'Time' and 'Amount'
legit = df[df.Class == 0]
fraud = df[df.Class == 1]

legit_sample = legit.sample(n=492, random_state=2) # Added random_state for reproducibility
new_dataset = pd.concat([legit_sample, fraud], axis=0)

# Shuffle the new dataset
new_dataset = new_dataset.sample(frac=1, random_state=2).reset_index(drop=True) # Added random_state for reproducibility and reset index

X = new_dataset.drop(columns='Class', axis=1)
y = new_dataset['Class']

# Split the data again with the scaled data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=2)

print("X after scaling and splitting:")
display(X.head())
print("\nX_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

X after scaling and splitting:


Time         V1        V2         V3         V4         V5        V6  \
0 -1.183726   1.214115  0.425012   0.632661   1.178179  -0.442638 -1.054988   
1 -1.580436 -14.179165  7.421370 -21.405836  11.927512  -7.974281 -2.202710   
2 -0.017875 -11.397727  7.763953 -18.572307   6.711855 -10.174216 -4.395918   
3  0.586382   2.016913 -1.219690  -1.055441  -0.567208  -0.676719 -0.149719   
4 -0.974579   1.243848  0.524526  -0.538884   1.209196   0.479538 -0.197429   

          V7        V8        V9  ...       V20       V21       V22       V23  \
0   0.182845 -0.265533 -0.122159  ... -0.087807 -0.021853  0.016541  0.010246   
1 -15.471612 -0.356595 -6.380125  ...  1.002764 -2.366836  1.130955  0.991153   
2 -15.893788  2.083013 -4.988837  ...  0.208879  0.339007  1.342923  0.239217   
3  -0.602194 -0.087608  0.322531  ...  0.108331 -0.125608 -0.269123  0.189139   
4   0.049166  0.037792  0.128119  ... -0.171541 -0.051660 -0.084089 -0.192846   

        V24       V25       V26       V27       V28    Amount  
0  0.733552  0.533972 -0.462462  0.029849  0.030895 -0.343234  
1 -1.033132 -0.327179  0.634693  2.171905 -1.395288 -0.349231  
2  0.534644 -0.174965 -0.500240 -1.722060 -0.574339 -0.307651  
3  0.628028 -0.060834 -0.355296 -0.017959 -0.039148  0.009597  
4 -0.917392  0.681953 -0.194419  0.045917  0.040136 -0.349231  

[5 rows x 30 columns]


X_train shape: (787, 30)
X_test shape: (197, 30)


In [94]:
#running training loop

torch.manual_seed(42)

epochs = 10000 # Increased number of epochs

for epoch in range(epochs):
  model.train()

  # Convert DataFrames to Tensors
  X_train_tensor = torch.from_numpy(X_train.values).float()
  y_train_tensor = torch.from_numpy(y_train.values).float()


  y_logits = model(X_train_tensor).squeeze()
  y_pred = torch.round(torch.sigmoid(y_logits))

  loss = loss_fn(y_logits, y_train_tensor)
  acc = accuracy_fn(y_true=y_train_tensor, y_pred=y_pred)

  optimizer.zero_grad()

  loss.backward()

  optimizer.step()

  model.eval()

  with torch.inference_mode():
    # Convert DataFrame to Tensor
    X_test_tensor = torch.from_numpy(X_test.values).float()
    y_test_tensor = torch.from_numpy(y_test.values).float()

    test_logits = model(X_test_tensor).squeeze()
    test_pred = torch.round(torch.sigmoid(test_logits))

    test_loss = loss_fn(test_logits, y_test_tensor)
    test_acc = accuracy_fn(y_true=y_test_tensor, y_pred=test_pred)

  if epoch % 500 == 0:
    print(f"Epoch: {epoch} | Loss: {loss:.5f}, Acc: {acc:.2f}% | Test Loss: {test_loss:.5f}, Test Acc: {test_acc:.2f}%")

Epoch: 0 | Loss: 0.68021, Acc: 69.00% | Test Loss: 0.42367, Test Acc: 76.65%
Epoch: 500 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.37462, Test Acc: 97.46%
Epoch: 1000 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.41793, Test Acc: 97.46%
Epoch: 1500 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.44608, Test Acc: 97.46%
Epoch: 2000 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.46701, Test Acc: 97.46%
Epoch: 2500 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.48465, Test Acc: 97.46%
Epoch: 3000 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.50027, Test Acc: 97.46%
Epoch: 3500 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.51433, Test Acc: 97.46%
Epoch: 4000 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.52736, Test Acc: 97.46%
Epoch: 4500 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.53986, Test Acc: 97.46%
Epoch: 5000 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.55097, Test Acc: 97.46%
Epoch: 5500 | Loss: 0.00000, Acc: 100.00% | Test Loss: 0.56211, Test Acc: 97.46%
Epoch: 6000 | Loss: 0.00000, Acc:

In [95]:
from sklearn.metrics import precision_score, recall_score, f1_score

# Set model to evaluation mode
model.eval()

# Make predictions on the test set
with torch.inference_mode():
    X_test_tensor = torch.from_numpy(X_test.values).float()
    y_test_tensor = torch.from_numpy(y_test.values).float()

    test_logits = model(X_test_tensor).squeeze()
    test_pred = torch.round(torch.sigmoid(test_logits)) # Changed back to rounding


# Convert predictions and ground truth to numpy arrays for scikit-learn metrics
y_true_np = y_test_tensor.numpy()
y_pred_np = test_pred.numpy()

# Calculate precision, recall, and F1-score
precision = precision_score(y_true_np, y_pred_np)
recall = recall_score(y_true_np, y_pred_np)
f1 = f1_score(y_true_np, y_pred_np)

print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

Precision: 0.9794
Recall: 0.9694
F1-score: 0.9744


# **Going to use entire data set, instead of the small segment**

In [96]:
X = df.drop(columns='Class', axis=1)
y = df['Class']

print(X.shape)
print(y.shape)

(284807, 30)
(284807,)


In [97]:
#splitting the dataset

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

print("Shapes after splitting the full dataset:")
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

Shapes after splitting the full dataset:
X_train shape: (227845, 30)
X_test shape: (56962, 30)
y_train shape: (227845,)
y_test shape: (56962,)


In [98]:
#calculating class weights and instantiating the loss function with these weights to address class imbalance.

from sklearn.utils.class_weight import compute_class_weight
import torch

# Calculate class weights for the imbalanced dataset
class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)

# Convert class weights to a PyTorch tensor
class_weights_tensor = torch.tensor(class_weights, dtype=torch.float)

# Instantiate the BCEWithLogitsLoss with class weights
# The pos_weight parameter should be the weight for the positive class (Class=1)
loss_fn = nn.BCEWithLogitsLoss(pos_weight=class_weights_tensor[1])

print(f"Class weights: {class_weights}")
print(f"Class weights tensor: {class_weights_tensor}")
print(f"Loss function with pos_weight: {loss_fn}")

Class weights: [  0.50086612 289.14340102]
Class weights tensor: tensor([  0.5009, 289.1434])
Loss function with pos_weight: BCEWithLogitsLoss()


In [99]:
#running training loop with class weights

torch.manual_seed(42)

epochs = 100 # Reduced number of epochs for demonstration on full dataset

for epoch in range(epochs):
  model.train()

  # Convert DataFrames to Tensors
  X_train_tensor = torch.from_numpy(X_train.values).float()
  y_train_tensor = torch.from_numpy(y_train.values).float()


  y_logits = model(X_train_tensor).squeeze()
  y_pred = torch.round(torch.sigmoid(y_logits))

  # Use the loss function with class weights
  loss = loss_fn(y_logits, y_train_tensor)
  acc = accuracy_fn(y_true=y_train_tensor, y_pred=y_pred)

  optimizer.zero_grad()

  loss.backward()

  optimizer.step()

  model.eval()

  with torch.inference_mode():
    # Convert DataFrame to Tensor
    X_test_tensor = torch.from_numpy(X_test.values).float()
    y_test_tensor = torch.from_numpy(y_test.values).float()

    test_logits = model(X_test_tensor).squeeze()
    test_pred = torch.round(torch.sigmoid(test_logits))

    # Use the loss function with class weights for testing as well
    test_loss = loss_fn(test_logits, y_test_tensor)
    test_acc = accuracy_fn(y_true=y_test_tensor, y_pred=test_pred)

  if epoch % 10 == 0: # Print more frequently for fewer epochs
    print(f"Epoch: {epoch} | Loss: {loss:.5f}, Acc: {acc:.2f}% | Test Loss: {test_loss:.5f}, Test Acc: {test_acc:.2f}%")

Epoch: 0 | Loss: 0.95442, Acc: 95.03% | Test Loss: 15.11920, Test Acc: 99.92%
Epoch: 10 | Loss: 0.32111, Acc: 99.94% | Test Loss: 6.43434, Test Acc: 99.83%
Epoch: 20 | Loss: 2.03403, Acc: 99.84% | Test Loss: 3.79737, Test Acc: 99.45%
Epoch: 30 | Loss: 1.87679, Acc: 98.80% | Test Loss: 2.94360, Test Acc: 99.53%
Epoch: 40 | Loss: 0.13235, Acc: 96.27% | Test Loss: 0.33028, Test Acc: 97.56%
Epoch: 50 | Loss: 0.06800, Acc: 98.87% | Test Loss: 0.29703, Test Acc: 98.98%
Epoch: 60 | Loss: 0.08822, Acc: 99.46% | Test Loss: 0.42078, Test Acc: 99.32%
Epoch: 70 | Loss: 13.94882, Acc: 99.56% | Test Loss: 0.44299, Test Acc: 99.28%
Epoch: 80 | Loss: 0.11301, Acc: 99.71% | Test Loss: 0.90838, Test Acc: 99.72%
Epoch: 90 | Loss: 0.01635, Acc: 99.76% | Test Loss: 2.42050, Test Acc: 99.72%


In [100]:
#evaluating the model

from sklearn.metrics import confusion_matrix

# Set model to evaluation mode
model.eval()

# Make predictions on the test set
with torch.inference_mode():
    X_test_tensor = torch.from_numpy(X_test.values).float()
    y_test_tensor = torch.from_numpy(y_test.values).float()

    test_logits = model(X_test_tensor).squeeze()
    test_pred = torch.round(torch.sigmoid(test_logits))

# Convert predictions and ground truth to numpy arrays for scikit-learn metrics
y_true_np = y_test_tensor.numpy()
y_pred_np = test_pred.numpy()

# Calculate precision, recall, and F1-score
precision = precision_score(y_true_np, y_pred_np)
recall = recall_score(y_true_np, y_pred_np)
f1 = f1_score(y_true_np, y_pred_np)

print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

# Calculate confusion matrix
cm = confusion_matrix(y_true_np, y_pred_np)
print("\nConfusion Matrix:")
print(cm)

Precision: 0.3789
Recall: 0.8776
F1-score: 0.5292

Confusion Matrix:
[[56723   141]
 [   12    86]]


## Summary:

### Data Analysis Key Findings

*   The full dataset consists of 284,807 samples and 30 features.
*   The dataset was split into training (80%, 227,845 samples) and testing (20%, 56,962 samples) sets, preserving the class distribution in both sets.
*   Class weights were calculated to address the imbalance, with a significantly higher weight assigned to the minority class (fraud).
*   The model was trained for 100 epochs using the BCEWithLogitsLoss with the positive class weight applied.
*   On the test set, the model achieved a Precision of 0.3789, a Recall of 0.8776, and an F1-score of 0.5292.
*   The confusion matrix shows 56723 True Negatives, 86 True Positives, 141 False Positives, and 12 False Negatives.

### Insights or Next Steps

*   While the high recall (0.8776) indicates the model is good at identifying fraudulent transactions, the lower precision (0.3789) suggests a significant number of legitimate transactions are being flagged as fraudulent. This trade-off between precision and recall is common in imbalanced datasets and fraud detection.
*   Explore further techniques to improve the balance between precision and recall, such as adjusting the prediction threshold, using different loss functions, or exploring alternative sampling methods (e.g., SMOTE) on the training data.
